consensus/state_test.go

```diff
@@ -1028,6 +1030,31 @@ func TestSetValidBlockOnDelayedPrevote(t *testing.T) {
 	assert.True(t, rs.ValidRound == round)
 }
 
+// regression for #2518
+func TestNoHearbeatWhenNotValidator(t *testing.T) {
+	cs, _ := randConsensusState(4)
+	cs.Validators = types.NewValidatorSet(nil) // make sure we are not in the validator set
+
+	cs.evsw.AddListenerForEvent("testing", types.EventProposalHeartbeat,
+		func(data tmevents.EventData) {
+			t.Errorf("Should not have broadcasted heartbeat")
+		})
+	go cs.proposalHeartbeat(10, 1)
+
+	cs.Stop()
+	time.Sleep(proposalHeartbeatIntervalSeconds * time.Second)
```

Comment:
why do we need this timeout?

Reply:
Though process behind this defensive:
If we don't wait at least `proposalHeartbeatIntervalSeconds`, the test method will terminate immediately after `go cs.proposalHeartbeat(10, 1)` and depending on performance,  a EventProposalHearbeat` would have been sent just after leaving the test method. And that's exactly what we want to ensure it doesn't happen.

Tried to clarify with a comment and added an extra second.